Source/WebCore/dom/Document.cpp

```diff
@@ -10363,7 +10377,7 @@ bool Document::hasElementWithPendingUserAgentShadowTreeUpdate(Element& element)
 
 bool Document::isSameSiteForCookies(const URL& url) const
 {
-    auto domain = isTopDocument() ? RegistrableDomain(securityOrigin().data()) : RegistrableDomain(siteForCookies());
+    auto domain = isTopDocumentBasedOnFrame() ? RegistrableDomain(securityOrigin().data()) : RegistrableDomain(siteForCookies());
```

Comment:
Why can't this just check isMainFrame() on m_frame?

Reply:
Good question. I think the issue is multiple `Document`s can point to the same `Frame` (`Document`'s frame is not changed after construction), but `Frame` only points to one `Document` (`Frame`'s document can change via `LocalFrame::setDocument`, and `Frame` does not uniquely own `Document`). 
The existing behavior of `isTopDocument()` is to return true only if the document is in use by the main frame, so I just kept the existing behavior for the new function.

Reply:
Why can't we just change the implementation of isTopDocument to do this main frame check for all callers?  I think that would be better, but the change log indicates there is breakage.  I'm interested to know more about the breakage.

Reply:
It turns out the breakage is caused by web process crash on assertion failure; the other places seem to work (not causing observable failure for http/tests/cookies tests), so I just update the patch as suggested.